### ISSUE TITLE
Issue with window moving/sizing due to animation timing

### ISSUE THREAD

**Issue by MarqueIV** [Hammerspoon/hammerspoon#3731]
There seems to be an issue with setFrame, the grid-related functions and anything else related to window moving/sizing, and I think I've identified the issue as being related to the new os animations.

Specifically, it seems if you attempt to move/size a window while there's an existing animation, that action gets interrupted with the new call, thus the window isn't moved/sized correctly.

The hack I've come up with to get around this is rather than this....
```
window.setFrame(targetFrame)
```
I now do this... 
```
window.setTopLeft({
    x=targetFrame.x,
    y=targetFrame.y
})
os.execute("sleep 0.25")
window.setSize({
    w=targetFrame.w,
    h=targetFrame.h
})

```
where 0.25 seconds is enough to ensure the move animation happens completely before the resize operation. (You can reverse the order and you'll have the same effect.)

I'm somewhat confident this is the issue as when I reduce the delay, say down to 0.1 seconds (any value less than the needed animation), windows don't end up where they're supposed to be.  It's also incredibly subtle if you're right at the limit of the animation as it can seem the sizing is only a few pixels off, when in reality, the correct values were set, it was just interrupted right before it was completely animated into place/size.

As I said, I have this workaround, but I'm not crazy about it as now there are essentially two animations that happen when moving/sizing windows, but at least they end up where they're supposed to be.

**Comment by MarqueIV** [comment]
Bueller? Bueller?

**Comment by minusfive** [comment]
Do you have an example of the issues you mention? Or perhaps can you share the code you're using? I use hs as my primary/only window manager and don't experience any issues, so curious. Perhaps there's something I can help spot?

**Comment by MarqueIV** [comment]
Can you try my above code without the delay and see if it moves as you expect?  You need to make sure the window moves a long distance or else it will appear to work correctly.

For instance, if you have a 4K screen, place the window off the lower-right of your screen, basically as far from the upper-left corner as you can get it.  Then use the move and size calls to move it to the upper left corner, maybe 1/8th the screen width and height.  (It's even easier if you have multiple monitors.)

The point is it has to move a long-enough distance to make sure the animations from the first call are *not* yet completed when the second call starts.  That's the issue, and why the pause 'fixes' it.

Lemme know if you can repro it that way.

**Comment by MarqueIV** [comment]
@minusfive, did you give it a try?

**Comment by minusfive** [comment]
@MarqueIV, no, sorry, I'm busy. But I did quickly record a video for you of things working fine for me with animation and without delay. I use a 5K screen (5125 x 2160):

https://github.com/user-attachments/assets/35cb32c1-0017-49a1-b5b6-b12e75ed9832

You can peruse my custom WindowManager spoon here, in case it helps: https://github.com/minusfive/dotfiles/blob/main/.config/hammerspoon/Spoons/WindowManager.spoon/init.lua

Top level I'd suggest trying `setFrameInScreenBounds` instead of `setFrame`.

**Comment by MarqueIV** [comment]
Just watched your video. I can see that working fine because it's not doing the resizing which is what causes the actual problem. Your video looks like it's only doing moves. Can you change it so it's resizing from a quarter to a half or full screen to centered that kind of thing.

I'll also try out your spoon when I'm back in front of my computer.

**Comment by minusfive** [comment]
It is resizing, I’m just doing it too quickly.

**Comment by minusfive** [comment]
@MarqueIV here's a new video:

https://github.com/user-attachments/assets/9ab98199-2f19-4d2e-92b3-f31216240477

**Comment by minusfive** [comment]
@MarqueIV did that help?

**Comment by MarqueIV** [comment]
Not sure if this makes a difference, but can you try toggling quickly between quarters and halves?  i.e. top-left, then bottom half, then right half, then bottom left, etc.  I do see yours are resizing, but only in one dimension.  Not sure if that's part of it.

**Comment by minusfive** [comment]
Won't be able to reply for a few days, but vertical movement/resizing isn't problematic in my experience either. Will try and record at some point when I have a chance.

You can tweak the animation speed with https://www.hammerspoon.org/docs/hs.window.html#animationDuration

**Comment by jannis-baum** [comment]
I'm also having the issue with `setFrame` and all its variations (`setFrameWithWorkarounds`, `setFrameInScreenBounds`) not being able to resize **and** set position. The only way I was able to get it to work was first running `setSize`, then waiting 0.25s as @MarqueIV suggested, and then running `setTopLeft`; the other way around it also never does both. On top of this, setting `hs.window.animationDuration = 0` does nothing, i.e. I'm stuck watching the 2 slow animations.

---

**My exact use case for context:** A function that resizes windows so they fit below [SketchyBar](https://github.com/FelixKratz/SketchyBar) on external screens, i.e. checking if the top window edge is in the top 32px of the screen, and if so moving the window down by the overlap and adjusting its height so the bottom edge stays in the same y-coordinate:

```lua
hs.window.animationDuration = 0 -- does nothing

local statusbar_height = 32
local function fit_win_below_sketchybar(window)
    local screen = window:screen()
    -- ignore built-in display, fits because of notch
    if string.find(screen:name(), 'Built%-in') then
        return
    end
    local win_frame = window:frame()
    -- compute overlap with sketchybar
    local overlap = screen:frame().y + statusbar_height - win_frame.y
    if overlap <= 0 then
        return
    end
    -- horrible hack to move & resize window
    -- https://github.com/Hammerspoon/hammerspoon/issues/3731
    window:setSize({ w = win_frame.w, h = win_frame.h - overlap })
    os.execute('sleep 0.25')
    window:setTopLeft({ x = win_frame.x, y = win_frame.y + overlap })
end
```

**Comment by omgreenfield** [comment]
@jannis-baum beautiful. Thank you for working through that and supplying a solution for everyone.

For an example use case, below is how I applied the workaround. All of the functions I've defined manipulate the focused window, so I adjusted it by:

1. Renaming `ax_enhanced_workaround` to `changeFocusedWindowDimensions` (my repo uses camelCase)
2. Adding the following to the top of the function:
   1. Get focused window
   2. Return if it's `nil`
   3. Get its application
3. Pass `window` into the `action` callback

The rest of it is the same.

```lua
local function changeFocusedWindowDimensions(action)
  local window = hs.window.focusedWindow()

  if not window then return end

  local app = window:application()
  local ax_app = hs.axuielement.applicationElement(app)

  -- original settings
  local was_enhanced = ax_app.AXEnhancedUserInterface
  local original_animation_duration = hs.window.animationDuration

  -- set & run action
  ax_app.AXEnhancedUserInterface = false
  hs.window.animationDuration = 0
  action(window)

  -- restore original settings
  hs.window.animationDuration = original_animation_duration
  ax_app.AXEnhancedUserInterface = was_enhanced
end
```

Then it can be used like this:

```lua
local function moveWindowToLeft()
  changeFocusedWindowDimensions(function (window)
    local frame = window:frame()
    local screen = window:screen()
    local max = screen:frame()
    frame.x = max.x
    frame.y = max.y
    frame.w = max.w / 2
    frame.h = max.h
    window:setFrame(frame, 0)
  end)
end
```
